common/src/environments.rs

```diff
@@ -0,0 +1,65 @@
+use crate::client::ClientWithApi;
+use crate::gadget::message::UserID;
+use crate::gadget::tangle::runtime::TangleRuntime;
+use crate::gadget::tangle::TangleEvent;
+use crate::prelude::{TangleProtocolMessage, TangleWorkManager};
+use crate::utils::serialize;
+use gadget_core::job_manager::WorkManagerInterface;
+use serde::Serialize;
+use sp_core::ecdsa;
+use std::error::Error;
+
+pub trait GadgetEnvironment {
+    type Event: Send + Sync + 'static;
+    type ProtocolMessage: Send + Sync + 'static;
+    type Client: ClientWithApi<Self::Event> + Send + Sync + 'static;
+    type JobManager: WorkManagerInterface;
+    type Error: Error + Send + Sync + 'static;
+
+    fn build_protocol_message<Payload: Serialize>(
+        associated_block_id: <Self::JobManager as WorkManagerInterface>::Clock,
+        associated_session_id: <Self::JobManager as WorkManagerInterface>::SessionID,
+        associated_retry_id: <Self::JobManager as WorkManagerInterface>::RetryID,
+        associated_task_id: <Self::JobManager as WorkManagerInterface>::TaskID,
+        from: UserID,
+        to: Option<UserID>,
+        payload: &Payload,
+        from_account_id: Option<ecdsa::Public>,
```

Comment:
I think we need to generalize the account doing the key/signing of stuff too. Give that power to the AVS developer.